src/CommandSet.hpp

```diff
@@ -0,0 +1,50 @@
+#pragma once
+
+#include <functional>
+#include <map>
+#include <string>
+#include <utility>
+#include "string_utils.hpp"
+
+template<typename T>
+class CommandSet {
+ public:
+  CommandSet() {}
+  
+  // add a command called 'name' with a value, which is available to characters
+  // at level 'level' or above.
+  void add(const std::string& name_in, T value, int level) {
+    auto name = lower_case(name_in);
+    commands_[name] = std::make_pair(value, level);
+  }
+  
+  std::optional<const T> get(const std::string& name_in, int level) const {
```

Comment:
So: here you're returning a copy of the T you were given. Making it `const` is just annoying for the caller: it's _their_ T now! so: I'd suggest `optional<T>` here.

SUPER SUPER ANNOYINGLY you can't have what you probably want, which is `optional<const T&>`. Debate rages in the C++ community over this. 

If you want to avoid a copy, then returning a `const T*` and `nullptr` instead is your best choice :(

Reply:
Oh, that works. I was expecting it to refuse to let me return a non-const T, because this is a const function. But I suppose as it's theoretically returning-by-value (despite the fact that T may be a pointer, reference or whatever) it makes sense to be able to return std::optional<T>.